### PR TITLE
Introduce tone-based surfaces and accent color add-ons - Part 2

### LINKED ISSUE #115912: Support tone-based surface and surface container `ColorScheme` roles
Guidelines: https://m3.material.io/styles/color/the-color-system/color-roles#0abbf8b7-61e1-49ee-9f97-4967beb1e4fe

image

New tone-based colors will replace the existing surface roles. The updated roles remove the use of opacity overlays and thus of surface tint colors. The updated roles are also no longer tied to elevation, offering more flexibility for large screens, and support for future feature development including global user-controlled contrast and color fidelity. Material recommends teams accept these updates across code, tokens, and figma.

The following diagram shows the recommended approach to remapping from the old to the new surfaces.

transition path

Google issue b/266528810 and Google-internal [Material Design rationale with FAQ](https://docs.google.com/document/d/14e08HJUGcB2qduQoq1iJYWhhuaINQr8ZnM5dMU6Pqgg/edit?resourcekey=0-bWD9xzOCfbq0yLiE24-Pnw)

### PR DESCRIPTION
**Pull request by QuncCccccc**
Fixes https://github.com/flutter/flutter/issues/115912

This PR is to introduce 19 new color roles and deprecate 3 color roles in `ColorScheme`.
**Tone-based surface colors** (7 colors): 
* surfaceBright
* surfaceDim
* surfaceContainer
* surfaceContainerLowest
* surfaceContainerLow
* surfaceContainerHigh
* surfaceContainerHighest

**Accent color add-ons** (12 colors):
* primary/secondary/tertiary-Fixed
* primary/secondary/tertiary-FixedDim
* onPrimary/onSecondary/onTertiary-Fixed
* onPrimary/onSecondary/onTertiary-FixedVariant

**Deprecated colors**:
* background -> replaced with surface
* onBackground -> replaced with onSurface
* surfaceVariant -> replaced with surfaceContainerHighest

Please checkout this [design doc](https://docs.google.com/document/d/1ODqivpM_6c490T4j5XIiWCDKo5YqHy78YEFqDm4S8h4/edit?usp=sharing) for more information:)

Screenshot 2024-01-08 at 4 56 51 PM

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Review by HansMuller — COMMENTED** [review]
I haven't read all the changes yet; just a few comments from yesterday.

File: packages/flutter/lib/src/material/color_scheme.dart [hunk 1]
```diff
@@ -116,17 +133,50 @@ class ColorScheme with Diagnosticable {
     Color? onInverseSurface,
     Color? inversePrimary,
     Color? surfaceTint,
```
[the remark this answers is not in the document]
Comment (QuncCccccc) [reply packages/flutter/lib/src/material/color_scheme.dart:135]: Even though `surfaceTint` is no longer used by any widgets, the token database still keep this color in the color scheme([carbon](https://carbon.googleplex.com/google-material-3/tokens/designSystems/20543ce18892f7d9/tokenSets/2904362cb1f205af?resourceName=designSystems/20543ce18892f7d9/tokenSets/2904362cb1f205af/tokens/331b4c485a7e9be5&tagName=designSystems/20543ce18892f7d9/contextTagGroups/382f29e657e86b8c/tags/215f4693b82ba1df&tagName=designSystems/20543ce18892f7d9/contextTagGroups/38657353fa5c2ab9/tags/0b5c40268cdb2499&tagName=designSystems/20543ce18892f7d9/contextTagGroups/10a8f197da4037f1/tags/3423009265c4d795)), so I didn't mark this as deprecated. Will add more explanation! Thanks a lot for taking time to review this big PR!

File: packages/flutter/lib/src/material/color_scheme.dart [hunk 2]
```diff
@@ -615,6 +934,25 @@ class ColorScheme with Diagnosticable {
   /// <https://www.w3.org/TR/UNDERSTANDING-WCAG20/visual-audio-contrast-contrast.html>.
   Color get onPrimaryContainer => _onPrimaryContainer ?? onPrimary;
 
+  final Color? _primaryFixed;
+  /// A color to use in places where [primaryContainer] is normally used but this color
+  /// stays the same color between light and dark themes.
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/color_scheme.dart:939]:
```suggestion
  /// A substitute for [primaryContainer] that's the same color for the dark and light themes.
```

File: packages/flutter/lib/src/material/color_scheme.dart [hunk 3]
```diff
@@ -639,6 +977,25 @@ class ColorScheme with Diagnosticable {
   /// <https://www.w3.org/TR/UNDERSTANDING-WCAG20/visual-audio-contrast-contrast.html>.
   Color get onSecondaryContainer => _onSecondaryContainer ?? onSecondary;
 
+  final Color? _secondaryFixed;
+  /// A color to use in places where [secondaryContainer] is normally used but this color
+  /// stays the same color between light and dark themes.
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/color_scheme.dart:982]:
```suggestion
  /// A substitute for [secondaryContainer] that's the same color for the dark and light themes.
```

File: packages/flutter/lib/src/material/color_scheme.dart [hunk 4]
```diff
@@ -666,6 +1023,25 @@ class ColorScheme with Diagnosticable {
   /// <https://www.w3.org/TR/UNDERSTANDING-WCAG20/visual-audio-contrast-contrast.html>.
   Color get onTertiaryContainer => _onTertiaryContainer ?? onTertiary;
 
+  final Color? _tertiaryFixed;
+  /// A color to use in places where [tertiaryContainer] is normally used but this color
+  /// stays the same color between light and dark themes.
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/color_scheme.dart:1028]:
```suggestion
  /// A substitute for [tertiaryContainer] that's the same color for dark and light themes.
```

File: packages/flutter/lib/src/material/color_scheme.dart [hunk 5]
```diff
@@ -713,8 +1079,51 @@ class ColorScheme with Diagnosticable {
...
+  /// A color that always stays the darkest in either dark or light theme.
+  Color get surfaceDim => _surfaceDim ?? surface;
+
+  final Color? _surfaceBright;
+  /// A color that always stays the lightest in either dark or light theme.
+  Color get surfaceBright => _surfaceBright ?? surface;
+
+  final Color? _surfaceContainerLowest;
+  /// A surface container color with the lightest tone. It is used to create the
+  /// least emphasis against the surface.
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/color_scheme.dart:1098]:
```suggestion
  /// A surface container color with the lightest tone and the least emphasis relative to the surface.
```

File: packages/flutter/lib/src/material/color_scheme.dart [hunk 6]
```diff
@@ -713,8 +1079,51 @@ class ColorScheme with Diagnosticable {
...
+    'This feature was deprecated after v3.18.0-0.1.pre.'
+  )
   Color get surfaceVariant => _surfaceVariant ?? surface;
 
+  final Color? _surfaceDim;
+  /// A color that always stays the darkest in either dark or light theme.
+  Color get surfaceDim => _surfaceDim ?? surface;
+
+  final Color? _surfaceBright;
+  /// A color that always stays the lightest in either dark or light theme.
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/color_scheme.dart:1093]:
```suggestion
  /// A color that's always the lightest in the dark or light theme.
```

File: packages/flutter/lib/src/material/color_scheme.dart [hunk 7]
```diff
@@ -713,8 +1079,51 @@ class ColorScheme with Diagnosticable {
...
   /// A color variant of [surface] that can be used for differentiation against
   /// a component using [surface].
+  @Deprecated(
+    'Use surfaceContainerHighest instead. '
+    'This feature was deprecated after v3.18.0-0.1.pre.'
+  )
   Color get surfaceVariant => _surfaceVariant ?? surface;
 
+  final Color? _surfaceDim;
+  /// A color that always stays the darkest in either dark or light theme.
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/color_scheme.dart:1089]:
```suggestion
  /// A color that's always darkest in the dark or light theme.
```

File: packages/flutter/lib/src/material/color_scheme.dart [hunk 8]
```diff
@@ -713,8 +1079,51 @@ class ColorScheme with Diagnosticable {
...
+  Color get surfaceBright => _surfaceBright ?? surface;
+
+  final Color? _surfaceContainerLowest;
+  /// A surface container color with the lightest tone. It is used to create the
+  /// least emphasis against the surface.
+  Color get surfaceContainerLowest => _surfaceContainerLowest ?? surface;
+
+  final Color? _surfaceContainerLow;
+  /// A surface container color with a lighter tone. It is used to create less
+  /// emphasis than [surfaceContainer] but more emphasis than [surfaceContainerLowest].
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/color_scheme.dart:1103]:
```suggestion
  /// A surface container color with a lighter tone that creates less
  /// emphasis than [surfaceContainer] but more emphasis than [surfaceContainerLowest].
```

File: packages/flutter/lib/src/material/color_scheme.dart [hunk 9]
```diff
@@ -713,8 +1079,51 @@ class ColorScheme with Diagnosticable {
...
+  Color get surfaceContainerLow => _surfaceContainerLow ?? surface;
+
+  final Color? _surfaceContainer;
+  /// A recommended color role for a contained area against the surface.
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/color_scheme.dart:1107]:
```suggestion
  /// A recommended color role for a distinct area within the surface.
```

File: packages/flutter/lib/src/material/color_scheme.dart [hunk 10]
```diff
@@ -713,8 +1079,51 @@ class ColorScheme with Diagnosticable {
...
+  ///
+  /// Surface container color roles are independent from elevation. They are used
+  /// to replace the opacity-based model which apply a tinted overlay on top of
+  /// surfaces based on different elevation levels.
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/color_scheme.dart:1111]:
```suggestion
  /// Surface container color roles are independent of elevation. They replace the old
  // opacity-based model which applied a tinted overlay on top of
  /// surfaces based on their elevation.
```

**Review by HansMuller — COMMENTED** [review]
This looks great. Just some small-time feedback about ColorScheme doc strings.

**Review by HansMuller — COMMENTED** [review]
The components with `defaultStyleOf` methods document their defaults. The documentation will need to be updated. There defaultStyleOf method definitions in:
```
menu_anchor.dart
text_button.dart
filled_button.dart
elevated_button.dart
outlined_button.dart
icon_button.dart
```
The problem I just noticed is that the defaults for `overlayColors` have changed. I assume that there are others.

**Comment by QuncCccccc** [comment]
> The components with `defaultStyleOf` methods document their defaults. The documentation will need to be updated. There defaultStyleOf method definitions in:
> 
> ```
> menu_anchor.dart
> text_button.dart
> filled_button.dart
> elevated_button.dart
> outlined_button.dart
> icon_button.dart
> ```
> 
> The problem I just noticed is that the defaults for `overlayColors` have changed. I assume that there are others.

Ah thanks for catching this! Fixing!

**Review by HansMuller — COMMENTED** [review]
These classes will also need to be sync'd up with the changes:

text_button.dart _TextButtonDefaultOverlay
filled_button.dart _FilledButtonDefaultOverlay
elevated_button.dart _ElevatedButtonDefaultOverlay
outlined_button.dart _OutlinedButtonDefaultOverlay
icon_button.dart _IconButtonDefaultOverlay
toggle_buttons.dart _ToggleButtonDefaultOverlay

File: packages/flutter/lib/src/material/color_scheme.dart [hunk 11]
```diff
@@ -37,21 +37,43 @@ import 'theme_data.dart';
...
+/// but they stay the same color between light and dark themes. The '-Dim' roles
+/// provides a stronger, more emphasized color with the same fixed behavior.
+///
 /// The remaining colors of the scheme are composed of neutral colors used for
 /// backgrounds and surfaces, as well as specific colors for errors, dividers
-/// and shadows.
+/// and shadows. Surface colors are used for backgrounds and large, low-emphasis
+/// areas of the screen.
+///
+/// Material 3 also introduced tone-based surfaces and surface containers.
```
Comment (guidezpl) [inline packages/flutter/lib/src/material/color_scheme.dart:51]:
```suggestion
/// Material 3 also introduces tone-based surfaces and surface containers.
```

**Review by guidezpl — APPROVED** [review]
Nice work!

File: packages/flutter/lib/src/material/bottom_app_bar.dart [hunk 12]
```diff
@@ -94,8 +94,8 @@ class BottomAppBar extends StatefulWidget {
   /// The bottom app bar's background color.
   ///
   /// If this property is null then [BottomAppBarTheme.color] of
-  /// [ThemeData.bottomAppBarTheme] is used. If that's null then
-  /// [ThemeData.bottomAppBarColor] is used.
+  /// [ThemeData.bottomAppBarTheme] is used. If that's null, the default value
+  /// is [ColorScheme.surfaceContainer].
```
[the remark this answers is not in the document]
Comment (QuncCccccc) [reply packages/flutter/lib/src/material/bottom_app_bar.dart]: Ah yes, this is the default value for material 3. I should mention the different default values in different condition(M2&M3)! Just fixed!

File: packages/flutter/lib/src/material/bottom_sheet.dart [hunk 13]
```diff
@@ -1403,16 +1402,16 @@ class _BottomSheetDefaultsM3 extends BottomSheetThemeData {
   late final ColorScheme _colors = Theme.of(context).colorScheme;
 
   @override
-  Color? get backgroundColor => _colors.surface;
+  Color? get backgroundColor => _colors.surfaceContainerLow;
 
   @override
-  Color? get surfaceTintColor => _colors.surfaceTint;
+  Color? get surfaceTintColor => null;
```
[the remark this answers is not in the document]
Comment (QuncCccccc) [reply packages/flutter/lib/src/material/bottom_sheet.dart]: Yes, should be `Colors.transparent`. Updated!

**Comment by QuncCccccc** [comment]
> _ToggleButtonDefaultOverlay

We didn't update toggle buttons to material 3. So except `ToggleButtons`, all other overlay color opacities are synced up with the latest M3 change:) Thanks!

File: examples/api/lib/material/color_scheme/color_scheme.0.dart [hunk 14]
```diff
@@ -0,0 +1,310 @@
...
+
+  @override
+  Widget build(BuildContext context) {
+    return Column(
+      children: <Widget>[
+        ColorGroup(children: <ColorChip>[
+          ColorChip(label: 'primary', color: colorScheme.primary, onColor: colorScheme.onPrimary),
+          ColorChip(label: 'onPrimary', color: colorScheme.onPrimary, onColor: colorScheme.primary),
+          ColorChip(
+              label: 'primaryContainer', color: colorScheme.primaryContainer, onColor: colorScheme.onPrimaryContainer),
```
Comment (HansMuller) [inline examples/api/lib/material/color_scheme/color_scheme.0.dart:135]: Two space indent here and below.  In this case, and in some of the others, just using one-line might be easier to read.

File: examples/api/lib/material/color_scheme/color_scheme.0.dart [hunk 15]
```diff
@@ -0,0 +1,310 @@
...
+  @override
+  Widget build(BuildContext context) {
+    return RepaintBoundary(
+      child: Card(clipBehavior: Clip.antiAlias, child: Column(children: children)),
+    );
+  }
+}
+
+class ColorChip extends StatelessWidget {
+  const ColorChip({
```
Comment (HansMuller) [inline examples/api/lib/material/color_scheme/color_scheme.0.dart:260]: Given the way this is used, a constructor with position arguments would be a little more compact.

File: examples/api/lib/material/color_scheme/color_scheme.0.dart [hunk 16]
```diff
@@ -0,0 +1,310 @@
...
+  });
+
+  final Color color;
+  final Color? onColor;
+  final String label;
+
+  static Color contrastColor(Color color) {
+    final Brightness brightness = ThemeData.estimateBrightnessForColor(color);
+    return switch (brightness) {
+    Brightness.dark  => Colors.white,
```
Comment (HansMuller) [inline examples/api/lib/material/color_scheme/color_scheme.0.dart:274]: two space indent

**Review by HansMuller — APPROVED** [review]
LGTM

**Comment by deadlyrazer** [comment]
Why do many of the colours use the fallback instead of implementing the colour? Having just the tokens and then adding it yourself doesn't make much sense when using the dynamic colour generator.

**Comment by QuncCccccc** [comment]
> Why do many of the colours use the fallback instead of implementing the colour? Having just the tokens and then adding it yourself doesn't make much sense when using the dynamic colour generator.

Hi! We've updated `ColorScheme.[fromSeed](https://github.com/flutter/flutter/blob/65cd84b58885fb78bd2d27f60fa7209bedf5be25/packages/flutter/lib/src/material/color_scheme.dart#L303)` and `[fromImageProvider](https://github.com/flutter/flutter/blob/65cd84b58885fb78bd2d27f60fa7209bedf5be25/packages/flutter/lib/src/material/color_scheme.dart#L1691)()` methods so they can build `ColorScheme`s where each color role will have their values from `MaterialDynamicColors`.

**Comment by deadlyrazer** [comment]
Wow that's great, thanks a lot!

**Comment by knott-todd** [comment]
Hey, I see this was merged to master. I'd really like to use it in my app, are there any plans to merge to beta? Flutter docs recommended against connecting to the master channel...

**Comment by guidezpl** [comment]
Betas are created monthly, so expect this change in the next beta release. Please see https://github.com/flutter/flutter/wiki/Where's-my-commit%3F#finding-the-versions-that-contain-framework-commit-x for more.

**Pull request merged**